File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/error/GreyError.scala
package com.grey.sdk.error

import io.grpc.{Status, StatusException, StatusRuntimeException}

/** Normalized error shape for all Grey SDK operations.
  *
  * @param code
  *   Error code identifying the type of error
  * @param message
  *   Human-readable error message
  * @param details
  *   Optional additional error details
  */
final case class GreyError(
    code: String,
    message: String,
    details: Option[Map[String, Any]] = None
) extends Exception(s"[$code] $message")

object GreyError:
  // Common error codes
  val Unauthorized: String = "UNAUTHORIZED"
  val Forbidden: String = "FORBIDDEN"
  val NotFound: String = "NOT_FOUND"
  val ValidationError: String = "VALIDATION_ERROR"
  val NetworkError: String = "NETWORK_ERROR"
  val Timeout: String = "TIMEOUT"
  val ServerError: String = "SERVER_ERROR"
  val Unknown: String = "UNKNOWN"

  /** Creates a GreyError from a gRPC StatusException. */
  def fromGrpcException(e: StatusException): GreyError =
    fromGrpcStatus(e.getStatus, Option(e.getTrailers).map(_.toString))

  /** Creates a GreyError from a gRPC StatusRuntimeException. */
  def fromGrpcRuntimeException(e: StatusRuntimeException): GreyError =
    fromGrpcStatus(e.getStatus, Option(e.getTrailers).map(_.toString))

  /** Creates a GreyError from a gRPC Status. */
  def fromGrpcStatus(status: Status, trailers: Option[String] = None): GreyError =
    val code = grpcStatusToCode(status.getCode)
    val message = Option(status.getDescription).getOrElse(status.getCode.name)
    val details = trailers.map(t => Map[String, Any]("grpcTrailers" -> t))
    GreyError(code, message, details)

  /** Creates a GreyError from any throwable. */
  def fromThrowable(e: Throwable): GreyError = e match
    case ge: GreyError              => ge
    case se: StatusException        => fromGrpcException(se)
    case sre: StatusRuntimeException => fromGrpcRuntimeException(sre)
    case other                      => GreyError(Unknown, other.getMessage)

  /** Maps gRPC status codes to Grey error codes. */
  private def grpcStatusToCode(grpcCode: Status.Code): String = grpcCode match
    case Status.Code.UNAUTHENTICATED    => Unauthorized
    case Status.Code.PERMISSION_DENIED  => Forbidden
    case Status.Code.NOT_FOUND          => NotFound
    case Status.Code.INVALID_ARGUMENT   => ValidationError
    case Status.Code.FAILED_PRECONDITION => ValidationError
    case Status.Code.UNAVAILABLE        => NetworkError
    case Status.Code.DEADLINE_EXCEEDED  => Timeout
    case Status.Code.INTERNAL           => ServerError
    case Status.Code.UNKNOWN            => Unknown
    case _                              => Unknown

  /** Creates an unauthorized error. */
  def unauthorized(message: String = "Not authenticated"): GreyError =
    GreyError(Unauthorized, message)

  /** Creates a validation error. */
  def validation(message: String): GreyError =
    GreyError(ValidationError, message)

/** Result type for operations that can fail.
  *
  * Provides a functional alternative to throwing exceptions.
  */
enum GreyResult[+A]:
  case Success(value: A)
  case Failure(error: GreyError)

  /** Whether this result is a success. */
  def isSuccess: Boolean = this match
    case Success(_) => true
    case Failure(_) => false

  /** Whether this result is a failure. */
  def isFailure: Boolean = !isSuccess

  /** Gets the value if success, otherwise throws. */
  def get: A = this match
    case Success(v) => v
    case Failure(e) => throw e

  /** Gets the value wrapped in Option. */
  def toOption: Option[A] = this match
    case Success(v) => Some(v)
    case Failure(_) => None

  /** Gets the error wrapped in Option. */
  def errorOption: Option[GreyError] = this match
    case Success(_) => None
    case Failure(e) => Some(e)

  /** Converts to Either. */
  def toEither: Either[GreyError, A] = this match
    case Success(v) => Right(v)
    case Failure(e) => Left(e)

  /** Maps the success value. */
  def map[B](f: A => B): GreyResult[B] = this match
    case Success(v) => Success(f(v))
    case Failure(e) => Failure(e)

  /** Flat maps the success value. */
  def flatMap[B](f: A => GreyResult[B]): GreyResult[B] = this match
    case Success(v) => f(v)
    case Failure(e) => Failure(e)

  /** Folds the result. */
  def fold[B](onFailure: GreyError => B, onSuccess: A => B): B = this match
    case Success(v) => onSuccess(v)
    case Failure(e) => onFailure(e)

  /** Recovers from a failure with a fallback value. */
  def recover[B >: A](pf: PartialFunction[GreyError, B]): GreyResult[B] = this match
    case Failure(e) if pf.isDefinedAt(e) => Success(pf(e))
    case other                           => other

object GreyResult:
  /** Creates a success result. */
  def success[A](value: A): GreyResult[A] = Success(value)

  /** Creates a failure result. */
  def failure[A](error: GreyError): GreyResult[A] = Failure(error)

  /** Creates a result from an Either. */
  def fromEither[A](either: Either[GreyError, A]): GreyResult[A] = either match
    case Right(v) => Success(v)
    case Left(e)  => Failure(e)

  /** Creates a result from a Try. */
  def fromTry[A](t: scala.util.Try[A]): GreyResult[A] = t match
    case scala.util.Success(v) => Success(v)
    case scala.util.Failure(e) => Failure(GreyError.fromThrowable(e))

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/error/grey_error.dart
/// Error handling types for Grey SDK.
library;

import 'package:grpc/grpc.dart';

/// Normalized error shape for all Grey SDK operations.
class GreyError implements Exception {
  /// Creates a new GreyError.
  const GreyError({
    required this.code,
    required this.message,
    this.details,
  });

  /// Creates a GreyError from a gRPC error.
  factory GreyError.fromGrpcError(GrpcError error) {
    final code = _grpcStatusToCode(error.code);
    return GreyError(
      code: code,
      message: error.message ?? 'Unknown gRPC error',
      details: <String, dynamic>{
        'grpcCode': error.code,
        'grpcCodeName': error.codeName,
        if (error.details != null) 'grpcDetails': error.details,
      },
    );
  }

  /// Creates a GreyError from a generic exception.
  factory GreyError.fromException(Object exception) {
    if (exception is GreyError) {
      return exception;
    }
    if (exception is GrpcError) {
      return GreyError.fromGrpcError(exception);
    }
    return GreyError(
      code: 'UNKNOWN',
      message: exception.toString(),
    );
  }

  /// Error code identifying the type of error.
  final String code;

  /// Human-readable error message.
  final String message;

  /// Optional additional error details.
  final Map<String, dynamic>? details;

  /// Common error codes.
  static const String unauthorized = 'UNAUTHORIZED';
  static const String forbidden = 'FORBIDDEN';
  static const String notFound = 'NOT_FOUND';
  static const String validationError = 'VALIDATION_ERROR';
  static const String networkError = 'NETWORK_ERROR';
  static const String timeout = 'TIMEOUT';
  static const String serverError = 'SERVER_ERROR';
  static const String unknown = 'UNKNOWN';

  @override
  String toString() => 'GreyError(code: $code, message: $message)';

  /// Copies this error with optional modifications.
  GreyError copyWith({
    String? code,
    String? message,
    Map<String, dynamic>? details,
  }) {
    return GreyError(
      code: code ?? this.code,
      message: message ?? this.message,
      details: details ?? this.details,
    );
  }
}

/// Maps gRPC status codes to Grey error codes.
String _grpcStatusToCode(int grpcCode) {
  return switch (grpcCode) {
    StatusCode.unauthenticated => GreyError.unauthorized,
    StatusCode.permissionDenied => GreyError.forbidden,
    StatusCode.notFound => GreyError.notFound,
    StatusCode.invalidArgument => GreyError.validationError,
    StatusCode.failedPrecondition => GreyError.validationError,
    StatusCode.unavailable => GreyError.networkError,
    StatusCode.deadlineExceeded => GreyError.timeout,
    StatusCode.internal => GreyError.serverError,
    StatusCode.unknown => GreyError.unknown,
    _ => GreyError.unknown,
  };
}

/// Result type for operations that can fail.
sealed class GreyResult<T> {
  const GreyResult._();

  /// Creates a success result.
  const factory GreyResult.success(T value) = GreySuccess<T>;

  /// Creates a failure result.
  const factory GreyResult.failure(GreyError error) = GreyFailure<T>;

  /// Whether this result is a success.
  bool get isSuccess => this is GreySuccess<T>;

  /// Whether this result is a failure.
  bool get isFailure => this is GreyFailure<T>;

  /// Gets the value if success, otherwise throws.
  T get value {
    return switch (this) {
      GreySuccess(:final value) => value,
      GreyFailure(:final error) => throw error,
    };
  }

  /// Gets the value if success, otherwise returns null.
  T? get valueOrNull {
    return switch (this) {
      GreySuccess(:final value) => value,
      GreyFailure() => null,
    };
  }

  /// Gets the error if failure, otherwise returns null.
  GreyError? get errorOrNull {
    return switch (this) {
      GreySuccess() => null,
      GreyFailure(:final error) => error,
    };
  }

  /// Maps the success value to a new type.
  GreyResult<U> map<U>(U Function(T value) transform) {
    return switch (this) {
      GreySuccess(:final value) => GreyResult.success(transform(value)),
      GreyFailure(:final error) => GreyResult.failure(error),
    };
  }

  /// Applies a function that returns a result to the success value.
  GreyResult<U> flatMap<U>(GreyResult<U> Function(T value) transform) {
    return switch (this) {
      GreySuccess(:final value) => transform(value),
      GreyFailure(:final error) => GreyResult.failure(error),
    };
  }

  /// Folds the result into a single value.
  U fold<U>({
    required U Function(T value) onSuccess,
    required U Function(GreyError error) onFailure,
  }) {
    return switch (this) {
      GreySuccess(:final value) => onSuccess(value),
      GreyFailure(:final error) => onFailure(error),
    };
  }
}

/// Success variant of GreyResult.
final class GreySuccess<T> extends GreyResult<T> {
  /// Creates a success result with the given value.
  const GreySuccess(this.value) : super._();

  /// The success value.
  @override
  final T value;
}

/// Failure variant of GreyResult.
final class GreyFailure<T> extends GreyResult<T> {
  /// Creates a failure result with the given error.
  const GreyFailure(this.error) : super._();

  /// The error that caused the failure.
  final GreyError error;
}
